File: internal/dns/dns.go
package dns

import (
	"fmt"
	"os/exec"
	"runtime"
)

func FlushDNS() error {
	switch runtime.GOOS {
	case "windows":
		return flushDNSWindows()
	case "darwin":
		return flushDNSDarwin()
	case "linux":
		return flushDNSLinux()
	default:
		return fmt.Errorf("unsupported operating system: %s", runtime.GOOS)
	}
}

func flushDNSWindows() error {
	cmd := exec.Command("ipconfig", "/flushdns")
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("error flushing DNS: %v, output: %s", err, string(output))
	}
	return nil
}

func flushDNSDarwin() error {
	// First flush the DNS cache
	cmd := exec.Command("dscacheutil", "-flushcache")
	if output, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("error flushing DNS cache: %v, output: %s", err, string(output))
	}

	// Then restart the mDNSResponder
	cmd = exec.Command("killall", "-HUP", "mDNSResponder")
	if output, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("error restarting mDNSResponder: %v, output: %s", err, string(output))
	}

	return nil
}

func flushDNSLinux() error {
	// Try different methods in order of preference
	methods := []struct {
		name string
		cmd  []string
	}{
		{"systemd-resolve", []string{"systemd-resolve", "--flush-caches"}},
		{"resolvectl", []string{"resolvectl", "flush-caches"}},
		{"service", []string{"service", "nscd", "restart"}},
		{"nscd", []string{"nscd", "-K"}},
		{"dnsmasq", []string{"systemctl", "restart", "dnsmasq"}},
		{"network-manager", []string{"systemctl", "restart", "NetworkManager"}},
		{"pdnsd", []string{"systemctl", "restart", "pdnsd"}},
	}

	var lastErr error
	for _, method := range methods {
		// Check if the command exists
		if _, err := exec.LookPath(method.cmd[0]); err == nil {
			cmd := exec.Command(method.cmd[0], method.cmd[1:]...)
			if output, err := cmd.CombinedOutput(); err == nil {
				return nil
			} else {
				lastErr = fmt.Errorf("error with %s: %v, output: %s", method.name, err, string(output))
			}
		}
	}

	// If none of the above methods worked, try a simple hosts file reload
	// This is a fallback that should work on most systems
	if err := reloadHostsFile(); err != nil {
		return fmt.Errorf("all DNS flush methods failed. Last error: %v", lastErr)
	}

	return nil
}

func reloadHostsFile() error {
	// Simple touch of hosts file to trigger reload
	cmd := exec.Command("touch", "/etc/hosts")
	if output, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("error reloading hosts file: %v, output: %s", err, string(output))
	}
	return nil
}

File: src/dns.rs
use anyhow::Result;
use std::process::Command;

pub async fn flush_dns() -> Result<()> {
    #[cfg(target_os = "windows")]
    {
        Command::new("ipconfig")
            .arg("/flushdns")
            .output()
            .context("Failed to flush DNS cache")?;
    }

    #[cfg(target_os = "macos")]
    {
        Command::new("dscacheutil")
            .arg("-flushcache")
            .output()
            .context("Failed to flush DNS cache")?;

        Command::new("killall")
            .args(["-HUP", "mDNSResponder"])
            .output()
            .context("Failed to restart mDNSResponder")?;
    }

    #[cfg(target_os = "linux")]
    {
        // Try to detect the init system
        let is_systemd = std::path::Path::new("/run/systemd/system").exists();

        if is_systemd {
            // Try systemd-specific methods first
            let systemd_methods = [
                ("systemd-resolve", &["--flush-caches"][..]),
                ("resolvectl", &["flush-caches"][..]),
            ];

            for (cmd, args) in systemd_methods {
                if let Ok(_) = Command::new(cmd).args(args).output() {
                    return Ok(());
                }
            }
        }

        // Try common DNS cache services
        let service_methods = [
            ("service", vec!["nscd", "restart"]),
            ("service", vec!["dnsmasq", "restart"]),
            ("service", vec!["named", "restart"]),
            ("systemctl", vec!["restart", "NetworkManager"]),
        ];

        for (cmd, args) in service_methods {
            if let Ok(output) = Command::new(cmd).args(&args).output() {
                if output.status.success() {
                    return Ok(());
                }
            }
        }

        // Try direct service control
        let direct_methods = [
            ("nscd", vec!["-K"]),
            ("killall", vec!["-HUP", "dnsmasq"]),
            ("rndc", vec!["flush"]),
        ];

        for (cmd, args) in direct_methods {
            if let Ok(output) = Command::new(cmd).args(&args).output() {
                if output.status.success() {
                    return Ok(());
                }
            }
        }

        // Method 1: Touch the hosts file
        let _ = Command::new("touch").arg("/etc/hosts").output();

        // Method 2: Notify NetworkManager if it exists
        if let Ok(_) = Command::new("nmcli").arg("general").arg("reload").output() {
            return Ok(());
        }
    }

    Ok(())
}
